### PR TITLE
ci: job for build and test on the CI

### PR DESCRIPTION
**Pull request by kuisathaverat**
* Support for creating jobs with JJBB
* Job to build and test on CI

related to https://github.com/elastic/elastic-agent-client/issues/6

### CODE REVIEW

File: .ci/Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    rateLimitBuilds(throttle: [count: 60, durationName: 'hour', userBoost: true])
+    quietPeriod(10)
+  }
+  triggers {
+    issueCommentTrigger('(?i).*(?:jenkins\\W+)?run\\W+(?:the\\W+)?tests(?:\\W+please)?.*')
+  }
+  stages {
+    stage('Checkout') {
+      steps {
+        deleteDir()
```
Comment (v1v) [inline .ci/Jenkinsfile:27]:
```suggestion
        pipelineManager([ cancelPreviousRunningBuilds: [ when: 'PR' ] ])
        deleteDir()
```

File: .ci/Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    durabilityHint('PERFORMANCE_OPTIMIZED')
+    rateLimitBuilds(throttle: [count: 60, durationName: 'hour', userBoost: true])
+    quietPeriod(10)
+  }
+  triggers {
+    issueCommentTrigger('(?i).*(?:jenkins\\W+)?run\\W+(?:the\\W+)?tests(?:\\W+please)?.*')
+  }
+  stages {
+    stage('Checkout') {
+      steps {
```
Comment (v1v) [inline .ci/Jenkinsfile:26]: NIT:
```suggestion
      options { skipDefaultCheckout() }
      steps {
```

File: .ci/Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+      agent {
+        dockerfile {
+          filename 'Dockerfile'
+          dir '.ci/jenkins-go-agent'
+        }
+      }
+      environment{
+        HOME = "${WORKSPACE}/${BASE_DIR}"
+      }
+      steps {
```
Comment (v1v) [inline .ci/Jenkinsfile:42]:
```suggestion
      options { skipDefaultCheckout() }
      steps {
```

File: .ci/Jenkinsfile [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
```
Comment (v1v) [inline .ci/Jenkinsfile:60]:
```suggestion
      options { skipDefaultCheckout() }
      steps {
```

File: .ci/Jenkinsfile [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
```
Comment (v1v) [inline .ci/Jenkinsfile:78]:
```suggestion
      options { skipDefaultCheckout() }
      steps {
```

File: .ci/Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        stash allowEmpty: true, name: 'source', useDefaultExcludes: false
+      }
+    }
+    stage('Lint') {
+      agent {
+        dockerfile {
+          filename 'Dockerfile'
+          dir '.ci/jenkins-go-agent'
+        }
+      }
```
Comment (v1v) [inline .ci/Jenkinsfile:38]: neat!

Comment (kuisathaverat) [reply .ci/Jenkinsfile:40]: I think you can use it with parallel stages 🤔could help to parallelize small tasks

**Comment by ph** [comment]
@kuisathaverat Should this PR tests be already run by Jenkins or we need some plumbing to do?

File: .ci/Jenkinsfile (hunk 6 above)
Comment (cachedout) [reply .ci/Jenkinsfile:40]: That is cool!

**Comment by kuisathaverat** [comment]
@ph until we merge it we do not have job in Jenkins so it does not run.

**Comment by ph** [comment]
@kuisathaverat changes looks good to me. I've checked the individual task.

File: .ci/Jenkinsfile [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+@Library('apm@current') _
+
+pipeline {
+  agent { label 'ubuntu && immutable' }
+  environment {
+    BASE_DIR="src/github.com/elastic/elastic-agent-client"
+    PIPELINE_LOG_LEVEL='INFO'
+  }
+  options {
+    timeout(time: 1, unit: 'HOURS')
```
Comment (blakerouse) [inline .ci/Jenkinsfile:12]: This is a very large timeout for a simple client, I would hate for it to get stuck for some reason and it just sit for 1 hour. I think like 15 mins or 30 mins at the most.

Comment (kuisathaverat) [reply .ci/Jenkinsfile:12]: It is a default value, this timeout is only used when something goes really wrong and the build stucks for some reason. We usually adjust this value to a more proper value after a few build on the CI.

**Comment by kuisathaverat** [comment]
@v1v we cannot use skipDefaultCheckout if we launch the agent from a Dockerfile, the file is not there :D

**Pull request merged**